rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/SemanticTokenizer.java

```diff
@@ -514,4 +538,71 @@ else if (Character.isSupplementaryCodePoint(currentChar)) {
             }
         }
     }
+
+    // The idea behind the patch is to dynamically map productions in Rascal's
+    // own grammar with no category (e.g, number literals), or with a legacy
+    // category (e.g., string literals), to current categories (i.e., semantic
+    // token types).
+    private static class RascalPatch {
+
+        // As an optimization, productions that have already been mapped in the
+        // past, are cached in an identity map/set for fast lookups in the
+        // future.
+        Map<IConstructor, String> positiveCache = new IdentityHashMap<>();
```

Comment:
Good idea, but we should use caffeine, and configure it to also clear the map. It's not a direct problem as we're only doing it for rascal productions, and they won't grow. But if we end up using for DSLs at some point, this will become a memory leak, as old productions (from a previous version of the grammar) will never be cleared.

We used to have cases where people would run out of memory after having eclipse open for a few weeks because of a cache like this.

Reply:
Fixed.

However, right now, I configured it so that cache entries are auto-evicted and garbage-collected as soon as the keys are no longer reachable. For Rascal's grammar, when the RascalServer runs for weeks, I expect this will never happen (because the grammar objects presumably are kept in memory the whole time). I could add other eviction policies based on time and/or number of cache entries, but I wonder if this is worth the payoff: given that Rascar's own parser is used all the time (when writing Rascal), many/most of the cache entries would very quickly be re-inserted I guess.